### PR TITLE
Add GitHub Copilot Chat Support

### PR DESCRIPTION
**Pull request by Hawkbawk**
# Summary

This commit implements Github Copilot Chat support within the existing Assistant panel/framework. It required a little bit of trickery and internal API modification, as Copilot doesn't use the same authentication-style as all of the existing providers, opting to use OAuth and a short lived API key instead of a straight API key. All existing Assistant features should work.

Release Notes:
- Added Github Copilot Chat support ([#4673](https://github.com/zed-industries/zed/issues/4673)).

## Screenshots
A screenshot showing a conversation between a user and Github Copilot Chat within the Zed editor.

### CODE REVIEW

**Comment by Hawkbawk** [comment]
@cla-bot check

File: crates/copilot/src/copilot_chat.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        response.body_mut().read_to_end(&mut body).await?;
+
+        let body_str = std::str::from_utf8(&body)?;
+
+        Err(anyhow!("Failed to request API token: {}", body_str))
+    }
+}
+
+pub async fn fetch_copilot_oauth_token() -> Result<String, anyhow::Error> {
+    let path = if cfg!(target_os = "linux") {
```
Comment (maxdeviant) [inline crates/copilot/src/copilot_chat.rs:160]: We should be using one of our well-known directory paths here instead of computing it on the fly: https://github.com/zed-industries/zed/blob/a5b8094082ec5c033e976572fb5cf48309e5d6b1/crates/paths/src/paths.rs

File: crates/copilot/src/copilot_chat.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+    match async_fs::metadata(&path).await {
+            Ok(_) => {
+                let contents = async_fs::read_to_string(&path).await?;
+
+                let hosts: serde_json::Value  = serde_json::from_str(&contents)?;
+
+                let host = hosts.get("github.com").ok_or_else(|| anyhow::anyhow!("No host found for github.com"))?.clone();
+
+                let token = host.get("oauth_token").ok_or_else(|| anyhow::anyhow!("No oauth token found for github.com"))?.as_str().unwrap();
```
Comment (maxdeviant) [inline crates/copilot/src/copilot_chat.rs:178]:
```suggestion
                let token = host.get("oauth_token").ok_or_else(|| anyhow::anyhow!("No OAuth token found for github.com"))?.as_str().unwrap();
```

File: crates/completion/src/copilot_chat.rs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        };
+
+        Self { _subscription: sub }
+    }
+}
+
+impl Render for AuthenticationPrompt {
+    fn render(&mut self, _cx: &mut ViewContext<Self>) -> impl IntoElement {
+        const LABEL: &str =
+            "To use the assistant panel or inline assistant, you must login to Github Copilot. Your Github account must have an active Copilot Chat subscription.";
```
Comment (maxdeviant) [inline crates/completion/src/copilot_chat.rs:305]: "GitHub" should always be spelled with an uppercase "H" in user-facing text:

```suggestion
            "To use the assistant panel or inline assistant, you must login to GitHub Copilot. Your GitHub account must have an active Copilot Chat subscription.";
```

File: crates/copilot/src/copilot_chat.rs (hunk 2 above)
Comment (Hawkbawk) [reply crates/copilot/src/copilot_chat.rs]: Fixed!

File: crates/completion/src/copilot_chat.rs (hunk 3 above)
Comment (Hawkbawk) [reply crates/completion/src/copilot_chat.rs]: Ah, I did not know that. Fixed it!

File: crates/copilot/src/copilot_chat.rs (hunk 1 above)
Comment (Hawkbawk) [reply crates/copilot/src/copilot_chat.rs]: I moved this logic into the paths package and simplified it a little bit at the same time. Let me know how it looks.

File: crates/paths/src/paths.rs [hunk 4]
```diff
@@ -212,6 +212,20 @@ pub fn copilot_dir() -> &'static PathBuf {
     COPILOT_DIR.get_or_init(|| support_dir().join("copilot"))
 }
 
+pub fn copilot_chat_config_path() -> &'static PathBuf {
+    static COPILOT_CHAT_CONFIG_DIR: OnceLock<PathBuf> = OnceLock::new();
+
+    COPILOT_CHAT_CONFIG_DIR.get_or_init(|| {
+        if cfg!(target_os = "windows") {
```
Comment (maxdeviant) [inline crates/paths/src/paths.rs:219]: We should use `config_dir()` here instead of inventing new directories to use.

```rs
COPILOT_CHAT_CONFIG_DIR.get_or_init(|| {
    config_dir().join("github-copilot").join("hosts.json")
})
```

**Comment by bennetbo** [comment]
Hey, just a heads up, we want to merge #14929 first which changes the way completion providers work. So this will have to wait until then, but I'm happy to review this in a couple of days.

**Comment by Hawkbawk** [comment]
@bennetbo That works for me. I looked through the recent pull requests a few days back and saw that one. It looks like some of the changes that you're making there might make this PR easier too, especially with regards to keeping the auth state and short-lived API tokens in sync with the provider model. I'll keep an eye out for when that gets merged and then rebase my work on top of it. Thanks for looking!

File: crates/paths/src/paths.rs (hunk 4 above)
Comment (Hawkbawk) [reply crates/paths/src/paths.rs:219]: The problem is that `config_dir()` points to Zed's config directory, while Copilot stores it's config in it's own directory, namely `~/.config/github-copilot`.

**Comment by Hawkbawk** [comment]
@bennetbo This should be in a good spot for review now. I might need some help writing tests, but I've also noticed none of the other language model providers have any, so I'm not sure what you want to do here about that. Let me know if you have any questions about the implementation.
This uses a similar approach as https://github.com/CopilotC-Nvim/CopilotChat.nvim, where we let the Copilot inline completion server generate an OAuth Token, fetch that token from the disk, get a short-lived API token, then use that to get completions from GitHub's API.

**Comment by bennetbo** [comment]
@Hawkbawk This is looking pretty good, thank you for putting in all the work. I did some minor refactoring and moved the api_token logic to the CopilotChat implementation, so that we do not need to expose this to the language model provider, ensuring that we always get a valid api token even when calling the CopilotChat methods from elsewhere. 

I am going to test this a bit more today to see if I find any hiccups.

**Comment by bennetbo** [comment]
Worked fine for me yesterday, I'll test it a bit more on nightly today and then it will be part of the preview release tomorrow. Thanks for working on this!

File: crates/paths/src/paths.rs (hunk 4 above)
Comment (ramenhost) [reply crates/paths/src/paths.rs:219]: I'm testing this on Windows, and the path appears to be `~\AppData\Local\github-copilot\hosts.json`. GitHub login in Assistant panel doesn't work because of this.

Comment (Hawkbawk) [reply crates/paths/src/paths.rs:219]: Hmmm, that's unfortunate. You'll likely want to raise an issue or create a pull request yourself so this gets fixed. This PR had already been merged. I can likely get to it in the next few days, though I don't have a Windows machine with me right now to test it properly.

**Comment by baldwindavid** [comment]
Thanks for all your work on this @Hawkbawk ! I hadn't used the Assistant up to this point, but looking forward to using it now that this has landed.

File: crates/paths/src/paths.rs (hunk 4 above)
Comment (ramenhost) [reply crates/paths/src/paths.rs:219]: The issue is resolved after correcting the path #15587. Great work on this!

**Comment by und3fined** [comment]
For fun, VSCode Copilot now use GPT-4o 

Screenshot 2024-08-02 at 12 27 28 AM
